internals/overlord/logstate/gatherer.go

```diff
@@ -0,0 +1,256 @@
+	// ensure the client is not blocking subsequent teardown steps.
+	clientCancel context.CancelFunc
+
+	pullers *pullerGroup
+	// All pullers send logs on this channel, received by main loop
+	entryCh chan servicelog.Entry
+}
+
+// logGathererArgs allows overriding the newLogClient method and time values
+// in testing.
+type logGathererArgs struct {
+	tickPeriod        time.Duration
+	timeoutFinalFlush time.Duration
+	// method to get a new client
+	newClient func(*plan.LogTarget) (logClient, error)
+}
+
+func newLogGatherer(target *plan.LogTarget) (*logGatherer, error) {
+	return newLogGathererInternal(target, logGathererArgs{})
+}
+
+// newLogGathererInternal contains the actual creation code for a logGatherer.
+// This function is used in the real implementation, but also allows overriding
+// certain configuration values for testing.
+func newLogGathererInternal(target *plan.LogTarget, args logGathererArgs) (*logGatherer, error) {
+	args = fillDefaultArgs(args)
+	client, err := args.newClient(target)
+	if err != nil {
+		return nil, fmt.Errorf("could not create log client: %w", err)
+	}
+
+	g := &logGatherer{
+		logGathererArgs: args,
+
+		targetName: target.Name,
+		client:     client,
+		entryCh:    make(chan servicelog.Entry),
+		pullers:    newPullerGroup(target.Name),
+	}
+	g.clientCtx, g.clientCancel = context.WithCancel(context.Background())
+	g.tomb.Go(g.loop)
+	return g, nil
+}
+
+func fillDefaultArgs(args logGathererArgs) logGathererArgs {
+	if args.tickPeriod == 0 {
+		args.tickPeriod = tickPeriod
+	}
+	if args.timeoutFinalFlush == 0 {
+		args.timeoutFinalFlush = timeoutFinalFlush
+	}
+	if args.newClient == nil {
+		args.newClient = newLogClient
+	}
+	return args
+}
+
+// planChanged is called by the LogManager when the plan is changed.
+func (g *logGatherer) planChanged(pl *plan.Plan, buffers map[string]*servicelog.RingBuffer) {
+	// Remove old pullers
+	g.pullers.RemoveOld(pl)
+
+	// Add new pullers
+	for _, service := range pl.Services {
+		target := pl.LogTargets[g.targetName]
+		if !service.LogsTo(target) {
+			continue
+		}
+
+		buffer, bufferExists := buffers[service.Name]
+		if !bufferExists {
+			// We don't yet have a reference to the service's ring buffer
+			// Need to wait until serviceStarted
+			continue
+		}
+
+		g.pullers.Add(service.Name, buffer, g.entryCh)
+	}
+}
+
+// serviceStarted is called by the LogManager on the start of a service which
+// logs to this gatherer's target.
+func (g *logGatherer) serviceStarted(service *plan.Service, buffer *servicelog.RingBuffer) {
+	g.pullers.Add(service.Name, buffer, g.entryCh)
+}
+
+// The main control loop for the logGatherer. loop receives logs from the
+// pullers on entryCh, and writes them to the client. It also flushes the
+// client periodically, and exits when the gatherer's tomb is killed.
+func (g *logGatherer) loop() error {
+	ticker := time.NewTicker(g.tickPeriod)
+	defer ticker.Stop()
+
+mainLoop:
```

Comment:
Can the second Flush block be incorporated in the main loop? Having two contexts feel a bit confusing to me. If the exit was not blocked by the client context (e.g. the pullers exited cleanly, fast), the context would still be valid, and we can use it to Flush on exit. If the first client context is dead, that is most likely because the client is not responding, so creating a second context feels slightly too hopeful.

Can we not simply have:
```
    flush := func() {
        err := g.client.Flush(g.clientCtx)
	if err != nil {
		logger.Noticef("sending logs to target %q: %v", g.targetName, err)
	}
	
	:
		select {
		case <-g.tomb.Dying():
			flush()
			break mainLoop
		case <-ticker.C:
		        flush()
	:
	
```

Reply:
> Can the second Flush block be incorporated in the main loop?

It could be, but I like having it separate. Makes it more clear that "this is the final step".

> Having two contexts feel a bit confusing to me.

There's a very good reason why we need this, although it is a little complex.

At the moment `gatherer.Stop()` is called, we may be in the middle of flushing the client (e.g. waiting for a response from the server). This could potentially take a long time, and while we're inside the `Flush` call, the main loop is blocked, and can't process any logs from the pullers. So we need to set a deadline, after which we cancel the flush, so we can unblock the loop and process those final logs.

So we need to create a new context for the final flush, as the previous one may have been cancelled. Plus, we want to have separate lifetimes for the current flush (if any) and the final flush.

> If the first client context is dead, that is most likely because the client is not responding, so creating a second context feels slightly too hopeful.

It could be the server not responding, or maybe our request got lost in the ether, due to a bad connection or something. In any case, I think we should still make a best effort to ship the final logs.